### PR TITLE
Integrate OAuth Web Worker library with identity apps

### LINKED ISSUE #1009: Improve token storage strategy to align with SPA best practices
**Description:**
Tokens and other sensitive data should not be vulnerable to XSS and  other attacks.

**Describe the solution you'd like**
Tokens and other security critical data should be stored in a secure manner. Better to look in to strategies like web workers or Javascript Closures.

### PR DESCRIPTION
**Pull request by thivi**
Fixes #1009

### CODE REVIEW

File: apps/admin-portal/src/api/certificates.ts [hunk 1]
```diff
@@ -18,14 +18,13 @@
...
-import { AxiosHttpClient, AxiosHttpClientInstance } from "@wso2is/http";
+import { OAuth } from "@wso2is/oauth-web-worker";
 import { store } from "../store";
 
 /**
  * Initialize an axios Http client.
- * @type { AxiosHttpClientInstance }
  */
-const httpClient: AxiosHttpClientInstance = AxiosHttpClient.getInstance();
+const httpClient = OAuth.getInstance().httpRequest;
```
Comment (brionmario) [inline apps/admin-portal/src/api/certificates.ts:27]: Is `httpRequest` a function in the instance returned by `OAuth` class?

File: apps/admin-portal/src/store/actions/authenticate.ts [hunk 2]
```diff
@@ -131,47 +129,42 @@ export const getProfileInformation = () => (dispatch): void => {
...
  * Handle user sign-in
  */
 export const handleSignIn = () => (dispatch) => {
-    identityManager.getInstance().signIn(
-        () => {
-            dispatch(setSignIn());
-            dispatch(getProfileInformation());
+    const oAuth = OAuth.getInstance();
+    oAuth
+        .initialize({
```
Comment (brionmario) [inline apps/admin-portal/src/store/actions/authenticate.ts:138]: How do we handle the response mode?

File: apps/user-portal/src/app.tsx [hunk 3]
```diff
@@ -67,7 +68,7 @@ export const App = (): ReactElement => {
         dispatch(setServiceResourceEndpoints<ServiceResourceEndpointsInterface>(Config.getServiceResourceEndpoints()));
         dispatch(setI18nConfigs<I18nModuleOptionsInterface>(Config.getI18nConfig()));
         dispatch(setUIConfigs<UIConfigInterface>(Config.getUIConfig()));
-    }, []);
+    }, [dispatch]);
```
Comment (brionmario) [inline apps/user-portal/src/app.tsx:71]: Shall we remove `dispatch` from the deps array. Even though we get an exhaustive dependency lint warning, adding functions in the array would be problematic and might need some extensive testing.

Check other places as well.

File: apps/user-portal/src/api/security-questions.ts [hunk 4]
```diff
@@ -58,16 +58,16 @@ export const getSecurityQs = (): Promise<any> => {
             url: store.getState().config.endpoints.challengeAnswers
         };
 
-        return httpClient.request(requestConfig);
+        return httpClient(requestConfig);
     };
 
-    return httpClient.all([ getQuestions(), getAnswers() ])
-        .then(httpClient.spread((questions, answers) => {
+    return Promise.all([ getQuestions(), getAnswers() ])
```
Comment (brionmario) [inline apps/user-portal/src/api/security-questions.ts:64]: Promise all might need some pollyfills to work in IE environments. Can't we bind `all` and `spread` to the worker instance?

File: apps/user-portal/src/utils/refresh-page-utils.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+import { history } from "../helpers";
```
Comment (brionmario) [inline apps/user-portal/src/utils/refresh-page-utils.ts:19]:
```suggestion

import { history } from "../helpers";
```

File: modules/oauth-web-worker/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
+module.exports = {
```
Comment (brionmario) [inline modules/oauth-web-worker/.eslintrc.js:1]: Shall we extend the existing common lint config.

```suggestion
module.exports = {
    extends: [
        "../../.eslintrc.js"
    ]
};

```

File: modules/oauth-web-worker/.eslintignore [hunk 7]
```diff
@@ -0,0 +1 @@
+dist/
```
Comment (brionmario) [inline modules/oauth-web-worker/.eslintignore:1]:
```suggestion
dist/
node_modules/
```

File: apps/admin-portal/src/components/certificates/certificates-list.tsx [hunk 8]
```diff
@@ -497,13 +498,15 @@ export const CertificatesList: FunctionComponent<CertificatesListPropsInterface>
...
+            && hasRequiredScopes(
+                featureConfig?.certificates, featureConfig?.certificates?.scopes?.read, allowedScopes
+            )) {
 
             return true;
         }
 
         return type === TRUSTSTORE
-            && hasRequiredScopes(featureConfig?.certificates, featureConfig?.certificates?.scopes?.read);
+            && hasRequiredScopes(featureConfig?.certificates, featureConfig?.certificates?.scopes?.read, allowedScopes);
```
[the remark this answers is not in the document]
Comment (thivi) [reply apps/admin-portal/src/components/certificates/certificates-list.tsx:509]: It's within the limit.

File: apps/admin-portal/src/api/certificates.ts (hunk 1 above)
Comment (thivi) [reply apps/admin-portal/src/api/certificates.ts:27]: Yes

File: apps/user-portal/src/api/security-questions.ts (hunk 4 above)
Comment (thivi) [reply apps/user-portal/src/api/security-questions.ts]: I added a new method called `httpRequestAll` to imitate the `axios.all` method.

File: apps/admin-portal/src/store/actions/authenticate.ts (hunk 2 above)
Comment (thivi) [reply apps/admin-portal/src/store/actions/authenticate.ts]: Added response mode to all initialize methods.

File: apps/admin-portal/src/api/certificates.ts (hunk 1 above)
Comment (brionmario) [reply apps/admin-portal/src/api/certificates.ts:27]: (y)

**Pull request merged**